File: python/task3/SurvivalGame.py
#
# CSCI3180 Principles of Programming Languages
#
# --- Declaration ---
#
# I declare that the assignment here submitted is original except for source
# material explicitly acknowledged. I also acknowledge that I am aware of
# University policy and regulations on honesty in academic work, and of the
# disciplinary guidelines and procedures applicable to breaches of such policy
# and regulations, as contained in the website
# http://www.cuhk.edu.hk/policy/academichonesty/
#
# Assignment 2
# Name : Wong Cho Hin
# Student ID : 1155070355
# Email Addr : wongch6@cse.cuhk.edu.hk
#
import random
from Pos import Pos
from Player import Player
from Chark import Chark
from Human import Human
from Obstacle import Obstacle
from Weapon import Weapon
from Axe import Axe
from Rifle import Rifle
class SurvivalGame:
    def __init__(self):
        #dimension of board
        self.D=10
        #number of obstacles
        self.O=2
        self.n=0
        self.teleportObjects=[]
        
    def printBoard(self):
        printObject=[[0 for i in range(self.D)]for j in range(self.D)]
        for i in range(0,self.D):
            for j in range(0,self.D):
                printObject[i][j]="  "
                
        for i in range(0,self.n):
            pos=self.teleportObjects[i].getPos()
            printObject[pos.getX()][pos.getY()]=self.teleportObjects[i].getName()
            
        for i in range(self.n,(self.n+self.O)):
            pos=self.teleportObjects[i].getPos()
            printObject[pos.getX()][pos.getY()]="O"+str(i-self.n)
        
        #printing
        string=" "
        for i in range(0,self.D):
            string+="| "+str(i)+"  "
        string+="|"
        print string
        
        string=""
        i=0
        while (i<self.D*5.5):
            string+="-"
            i+=1
        print string
        
        for row in range(0,self.D):
            string=""
            string+=str(row)
            for col in range(0,self.D):
                string+="| "+printObject[row][col]+" "
            string+="|"
            print string
            string=""
            i=0
            while (i<self.D*5.5):
                string+="-"
                i+=1
            print string
        
    def positionOccupied(self,randx,randy):
        for obj in self.teleportObjects:
            pos=obj.getPos()
            if (pos.getX()==randx)and(pos.getY()==randy):
                return True
            
        return False
    
    def getPlayer(self,randx,randy):
        for obj in self.teleportObjects:
            if(isinstance(obj,Player)):
                pos=obj.getPos()
                if(pos.getX()==randx)and(pos.getY()==randy):
                    return obj
        return None
        
    #---    
    def init(self):
        print "Welcome to Kafustrok. Light blesses you. \nInput number of players: (a even number)"
        self.n=input()
        self.teleportObjects=[0 for x in range(self.n+self.O)]
        #create N/2 Humans
        i=0
        while i<(self.n/2):
            self.teleportObjects[i]=Human(0,0,i,self)
            self.teleportObjects[i+self.n/2]=Chark(0,0,i,self)
            i+=1
        #create O obstacles. You cannot stand there   
        i=0
        while i<self.O:
            self.teleportObjects[i+self.n]=Obstacle(0,0,i,self)
            i+=1
        
    def gameStart(self):
        turn=0
        numOfAlivePlayers = self.n
        while numOfAlivePlayers>1:
            if turn==0:
                for obj in self.teleportObjects:
                    obj.teleport()
                print "Everything gets teleported.."
            self.printBoard()
            t=self.teleportObjects[turn]
            if(t.health>0):
                t.askForMove()
                print ""
            turn = (turn+1)%self.n
            numOfAlivePlayers=0
            for i in range(0,self.n):
                if(self.teleportObjects[i].health>0):
                    numOfAlivePlayers+=1
        print "Game over."
        self.printBoard()
            

        
        
game=SurvivalGame()
game.init()
game.gameStart()

File: java/SurvivalGame.java
import java.util.Scanner;

public class SurvivalGame {
	private int n; // Number of player
	public final int D = 10; // dimension of board
	private final int O = 2; // Number of obstacles

	
	private Object[] teleportObjects;

	public static Scanner reader = new Scanner(System.in);
    
    public int getN(){
        return this.n;
    }

	public void printBoard() {
		String printObject[][] = new String[D][D];

		// init printObject
		for (int i = 0; i < D; i++)
			for (int j = 0; j < D; j++)
				printObject[i][j] = "  ";

		for (int i = 0; i < n; i++) {
			Pos pos = ((Player)teleportObjects[i]).getPos();
			printObject[pos.getX()][pos.getY()] = ((Player) teleportObjects[i]).getName();
		}

		for (int i = n; i < n+O; i++) {
			Pos pos = ((Obstacle)teleportObjects[i]).getPos();
			printObject[pos.getX()][pos.getY()] = "O" + Integer.toString(i-n);
		}

		// printing
		System.out.print(" ");
		for (int i = 0; i < D; i++)
			System.out.print(String.format("| %d  ", i));

		System.out.println("|");

		for (int i = 0; i < D * 5.5; i++)
			System.out.print("-");
		System.out.println("");

		for (int row = 0; row < D; row++) {
			System.out.print(row);
			for (int col = 0; col < D; col++)
				System.out.print(String.format("| %s ",
				printObject[row][col]));
			System.out.println("|");
			for (int i = 0; i < D * 5.5; i++)
				System.out.print("-");
			System.out.println("");
		}

	}

	public boolean positionOccupied(int randx, int randy) {

		for (Object o : teleportObjects) {
			if (o instanceof Player) {
				Pos pos = ((Player) o).getPos();
				if (pos.getX() == randx && pos.getY() == randy)
					return true;
			} else {
				Pos pos = ((Obstacle) o).getPos();
				if (pos.getX() == randx && pos.getY() == randy)
					return true;
			}

		}

		return false;
	}

	public Player getPlayer(int randx, int randy) {
		// TODO Auto-generated method stub
		for (Object o : teleportObjects) {
			if (o instanceof Player) {
				Pos pos = ((Player) o).getPos();
				if (pos.getX() == randx && pos.getY() == randy)
					return (Player) o;
			}
		}

		return null;
	}

	private  void init() {

		System.out.println("Welcome to Kafustrok. Light blesses you. \nInput number of players: (a even number)");
		n = reader.nextInt();

		teleportObjects = new Object[n + O];

		// create N/2 Humans

		for (int i = 0; i < n / 2; i++) {
			teleportObjects[i] = new Human(0, 0, i, this);
			teleportObjects[i + n / 2] = new Chark(0, 0, i, this);
		}

		// create O obstacles. You cannot stand there
		for (int i = 0; i < O; i++) {

			teleportObjects[i + n] = new Obstacle(0, 0, i, this);
		}

		// positions would be reinitialized later. 0,0 is dummy

	}

	private  void gameStart() {
		int turn = 0;
		int numOfAlivePlayers = n;
        boolean end=false;
        String getClass="";
		while (end!=true) {
			// teleport after every N turns
            end=true;
			if (turn == 0) {
				for (Object obj : teleportObjects) {
					if (obj instanceof Human)
						((Human) obj).teleport();
					else if (obj instanceof Chark)
						((Chark) obj).teleport();
					else if (obj instanceof Obstacle)
						((Obstacle) obj).teleport();
				}
				System.out.println("Everything gets teleported..");
			}
			//printBoard();
			Player t = (Player) teleportObjects[turn];
			// t can move only if he is alive!
			if (t.health > 0) {
				// dynamic binding helps
                printBoard();
				t.askForMove();
				System.out.println("\n");
				
			}
			turn = (turn + 1) % n;
			// count number of alive players
			numOfAlivePlayers = 0;
            for (Object obj : teleportObjects) {
                if(!(obj instanceof Obstacle)){
                    if(((Player)obj).getHealth()>0){
                        getClass=obj.getClass().getSimpleName();
                        break;
                    }
                }
            }
            for (Object obj : teleportObjects) {
                if(!(obj instanceof Obstacle)){
                    if(((Player)obj).getHealth()>0){
                        if(!(obj.getClass().getSimpleName().equals(getClass))){
                            end=false;
                            break;
                        }
                    }
                }
            }

			for (int i = 0; i < n; i++) {
				if (((Player) teleportObjects[i]).health > 0)
					numOfAlivePlayers += 1;
			}

		}

		System.out.println("Game over.");
		printBoard();

	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		// System.out.println(String.format("(%d,%d)", 3,4));
		SurvivalGame game = new SurvivalGame();
		game.init();
		game.gameStart();
	}
}
